### PR TITLE
support fail fast for node removal

### PR DESCRIPTION
**Pull request by car-roll**
https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/372 added fail fast for cloud node removal. However, this feature [is not supported](https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/372#discussion_r1795825003) by the `ec2` plugin.

This PR refactors `EC2AbstractSlave` and `EC2Computer` in order to take advantage of https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/372 

### Testing done

- [ ] Perform manual testing by killing spot instance manually and watching startup

### Submitter checklist
- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/test/java/hudson/plugins/ec2/EC2RetentionStrategyTest.java
Comment (jglick) [inline src/test/java/hudson/plugins/ec2/EC2RetentionStrategyTest.java:1]: Failing in CI.

File: src/main/java/hudson/plugins/ec2/EC2AbstractSlave.java [hunk 1]
```diff
@@ -526,15 +523,23 @@ public boolean isAcceptingTasks() {
     void idleTimeout() {
         LOGGER.info("EC2 instance idle time expired: " + getInstanceId());
         if (!stopOnTerminate) {
-            terminate();
+            try {
+                terminate();
```
Comment (jglick) [inline src/main/java/hudson/plugins/ec2/EC2AbstractSlave.java:527]: (ignore WS)

File: src/main/java/hudson/plugins/ec2/EC2OndemandSlave.java [hunk 2]
```diff
@@ -109,6 +111,7 @@ public void terminate() {
                             Jenkins.get().removeNode(this);
                             LOGGER.info("Removed EC2 instance from jenkins controller: " + getInstanceId());
                         } catch (AmazonClientException | IOException e) {
+                            listener.error("Failed to terminate EC2 instance: " + getInstanceId());
                             LOGGER.log(Level.WARNING, "Failed to terminate EC2 instance: " + getInstanceId(), e);
```
Comment (jglick) [inline src/main/java/hudson/plugins/ec2/EC2OndemandSlave.java:115]: Could perhaps remove the system log message and send the stack trace to the `listener` (see `Functions.printStackTrace`).

File: src/test/java/hudson/plugins/ec2/EC2AbstractSlaveTest.java [hunk 3]
```diff
@@ -29,9 +32,8 @@ public void testGetLaunchTimeoutInMillisShouldNotOverflow() throws Exception {
             DEFAULT_METADATA_ENDPOINT_ENABLED, DEFAULT_METADATA_TOKENS_REQUIRED, DEFAULT_METADATA_HOPS_LIMIT, DEFAULT_METADATA_SUPPORTED) {
 
             @Override
-            public void terminate() {
-                // To change body of implemented methods use File | Settings |
-                // File Templates.
+            protected void _terminate(TaskListener listener) throws IOException, InterruptedException {
+                return;
```
Comment (jglick) [inline src/test/java/hudson/plugins/ec2/EC2AbstractSlaveTest.java:36]:
```suggestion
```

File: src/test/java/hudson/plugins/ec2/EC2AbstractSlaveTest.java [hunk 4]
```diff
@@ -56,7 +58,8 @@ public void testMaxUsesBackwardCompat() throws Exception {
         r.jenkins.clouds.add(ac);
         EC2AbstractSlave slave = new EC2AbstractSlave("name", "", description, "fs", 1, null, "label", null, null, "init", "tmpDir", new ArrayList<NodeProperty<?>>(), "root", "jvm", false, "idle", null, cloudName, false, Integer.MAX_VALUE, new UnixData("remote", null, null, "22", null), ConnectionStrategy.PRIVATE_IP, 0)  {
             @Override
-            public void terminate() {
+            protected void _terminate(TaskListener listener) throws IOException, InterruptedException {
+                return;
```
Comment (jglick) [inline src/test/java/hudson/plugins/ec2/EC2AbstractSlaveTest.java:62]:
```suggestion
```

File: src/test/java/hudson/plugins/ec2/EC2RetentionStrategyTest.java [hunk 5]
```diff
@@ -210,7 +211,8 @@ private EC2Computer computerWithIdleTime(final int minutes, final int seconds) t
     private EC2Computer computerWithIdleTime(final int minutes, final int seconds, final Boolean isOffline, final Boolean isConnecting) throws Exception {
         final EC2AbstractSlave slave = new EC2AbstractSlave("name", "id", "description", "fs", 1, null, "label", null, null, "init", "tmpDir", new ArrayList<NodeProperty<?>>(), "remote", "jvm", false, "idle", null, "cloud", false, Integer.MAX_VALUE, null, ConnectionStrategy.PRIVATE_IP, -1) {
             @Override
-            public void terminate() {
+            protected void _terminate(TaskListener listener) throws IOException, InterruptedException {
+                return;
```
Comment (jglick) [inline src/test/java/hudson/plugins/ec2/EC2RetentionStrategyTest.java:215]:
```suggestion
```

File: src/test/java/hudson/plugins/ec2/EC2RetentionStrategyTest.java [hunk 6]
```diff
@@ -279,7 +281,8 @@ private EC2Computer computerWithUpTime(final int minutes, final int seconds, fin
         idleTimeoutCalled.set(false);
         final EC2AbstractSlave slave = new EC2AbstractSlave("name", "id", "description", "fs", 1, null, "label", null, null, "init", "tmpDir", new ArrayList<NodeProperty<?>>(), "remote", "jvm", false, "idle", null, "cloud", false, Integer.MAX_VALUE, null, ConnectionStrategy.PRIVATE_IP, -1) {
             @Override
-            public void terminate() {
+            protected void _terminate(TaskListener listener) throws IOException, InterruptedException {
+                return;
```
Comment (jglick) [inline src/test/java/hudson/plugins/ec2/EC2RetentionStrategyTest.java:285]:
```suggestion
```

File: src/test/java/hudson/plugins/ec2/ssh/verifiers/SshHostKeyVerificationStrategyTest.java [hunk 7]
```diff
@@ -339,7 +340,8 @@ public MockEC2Computer(EC2AbstractSlave slave) {
         private static MockEC2Computer createComputer(String suffix) throws Exception {
             final EC2AbstractSlave slave = new EC2AbstractSlave(COMPUTER_NAME + suffix, "id" + suffix, "description" + suffix, "fs", 1, null, "label", null, null, "init", "tmpDir", new ArrayList<NodeProperty<?>>(), "remote", "jvm", false, "idle", null, "cloud", false, Integer.MAX_VALUE, null, ConnectionStrategy.PRIVATE_IP, -1) {
                 @Override
-                public void terminate() {
+                protected void _terminate(TaskListener listener) throws IOException, InterruptedException {
+                    return;
```
Comment (jglick) [inline src/test/java/hudson/plugins/ec2/ssh/verifiers/SshHostKeyVerificationStrategyTest.java:344]:
```suggestion
```

**Comment by jglick** [comment]
Looks right, but has it been tested?

File: src/test/java/hudson/plugins/ec2/EC2RetentionStrategyTest.java
Comment (car-roll) [reply src/test/java/hudson/plugins/ec2/EC2RetentionStrategyTest.java:1]: Yeah, this one is mystifying me. I'm still trying to figure out why

**Comment by car-roll** [comment]
`Cannot read field "terminateScheduled" because "node" is null`
Comparing the test logs from `master` and this PR, everything looks similar up until the very end. Both versions show the ec2 instance idle timeout and terminated/removed, but then this version has this error that the node is null. I guess something funky is going on with this version's terminate. Terminated too soon maybe?

File: src/main/java/hudson/plugins/ec2/EC2OndemandSlave.java (hunk 2 above)
Comment (car-roll) [reply src/main/java/hudson/plugins/ec2/EC2OndemandSlave.java]: fixed in 8b95a323909b18a7bf4bba82fe92ad942520dfec

File: src/main/java/hudson/plugins/ec2/EC2OndemandSlave.java [hunk 8]
```diff
@@ -111,8 +111,7 @@ protected void _terminate(TaskListener listener) throws IOException, Interrupted
                             Jenkins.get().removeNode(this);
                             LOGGER.info("Removed EC2 instance from jenkins controller: " + getInstanceId());
                         } catch (AmazonClientException | IOException e) {
-                            listener.error("Failed to terminate EC2 instance: " + getInstanceId());
-                            LOGGER.log(Level.WARNING, "Failed to terminate EC2 instance: " + getInstanceId(), e);
+                            e.printStackTrace(listener.error("Failed to terminate EC2 instance: " + getInstanceId()));
```
[the remark this answers is not in the document]
Comment (car-roll) [reply src/main/java/hudson/plugins/ec2/EC2OndemandSlave.java]: fixed in 8b95a323909b18a7bf4bba82fe92ad942520dfec

File: src/main/java/hudson/plugins/ec2/EC2SpotSlave.java [hunk 9]
```diff
@@ -84,8 +84,7 @@ protected void _terminate(TaskListener listener) throws IOException, Interrupted
                                 LOGGER.info("Cancelled Spot request: " + spotInstanceRequestId);
                             } catch (AmazonClientException e) {
                                 // Spot request is no longer valid
-                                listener.error("Failed to cancel Spot request: " + spotInstanceRequestId);
-                                LOGGER.log(Level.WARNING, "Failed to cancel Spot request: " + spotInstanceRequestId, e);
+                                e.printStackTrace(listener.error("Failed to cancel Spot request: " + spotInstanceRequestId));
```
[the remark this answers is not in the document]
Comment (car-roll) [reply src/main/java/hudson/plugins/ec2/EC2SpotSlave.java]: fixed in 8b95a323909b18a7bf4bba82fe92ad942520dfec

File: src/main/java/hudson/plugins/ec2/EC2SpotSlave.java [hunk 10]
```diff
@@ -102,8 +101,7 @@ protected void _terminate(TaskListener listener) throws IOException, Interrupted
                                         LOGGER.info("Terminated EC2 instance (terminated): " + instanceId);
                                     } catch (AmazonClientException e) {
                                         // Spot request is no longer valid
-                                        listener.error("Failed to terminate the Spot instance: " + instanceId);
-                                        LOGGER.log(Level.WARNING, "Failed to terminate the Spot instance: " + instanceId, e);
+                                        e.printStackTrace(listener.error("Failed to terminate the Spot instance: " + instanceId));
```
[the remark this answers is not in the document]
Comment (car-roll) [reply src/main/java/hudson/plugins/ec2/EC2SpotSlave.java]: fixed in 8b95a323909b18a7bf4bba82fe92ad942520dfec

File: src/main/java/hudson/plugins/ec2/EC2SpotSlave.java [hunk 11]
```diff
@@ -117,8 +115,7 @@ protected void _terminate(TaskListener listener) throws IOException, Interrupted
                             try {
                                 Jenkins.get().removeNode(this);
                             } catch (IOException e) {
-                                listener.error("Failed to remove agent");
-                                LOGGER.log(Level.WARNING, "Failed to remove agent: " + name, e);
+                                e.printStackTrace(listener.error("Failed to remove agent"));
```
[the remark this answers is not in the document]
Comment (car-roll) [reply src/main/java/hudson/plugins/ec2/EC2SpotSlave.java]: fixed in 8b95a323909b18a7bf4bba82fe92ad942520dfec

**Comment by car-roll** [comment]
closing in favor of #1015 due to the branch conflicts

**Comment by car-roll** [comment]
@jglick yeah I was having a terrible time with the merge conflicts but I didn't think about force pushing. Will do that next time.

**Pull request closed without merging**